src/NeoSharp.Core/Messaging/Handlers/BlockHeadersMessageHandler.cs

```diff
@@ -1,33 +1,84 @@
     {
         private readonly IBlockchain _blockchain;
+        private readonly IBinarySerializer _serializier;
+        private readonly ICrypto _crypto;
         private readonly ILogger<BlockHeadersMessageHandler> _logger;
 
-        public BlockHeadersMessageHandler(IBlockchain blockchain, ILogger<BlockHeadersMessageHandler> logger)
+        public BlockHeadersMessageHandler(IBlockchain blockchain, ILogger<BlockHeadersMessageHandler> logger, IBinarySerializer serializier, ICrypto crypto)
         {
             _blockchain = blockchain ?? throw new ArgumentNullException(nameof(blockchain));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _serializier = serializier ?? throw new ArgumentNullException(nameof(serializier));
+            _crypto = crypto ?? throw new ArgumentNullException(nameof(crypto));
         }
 
         public async Task Handle(BlockHeadersMessage message, IPeer sender)
         {
-            _blockchain.AddBlockHeaders(message.Payload.Headers.Select(u => u.Header));
+            var indexB = _blockchain.CurrentBlock.Index;
+            var indexH = _blockchain.LastBlockHeader.Index;
 
-            if (_blockchain.LastBlockHeader.Index < sender.Version.CurrentBlockIndex)
+            var notFoundBlocks = new List<UInt256>();
+            var notFoundHeaders = new List<BlockHeaderBase>();
+
+            foreach (var header in message.Payload.Headers)
+            {
+                var needHeader = indexH < header.Header.Index;
+                var needBlock = needHeader || indexB < header.Header.Index;
+
+                if (needBlock || needHeader)
+                {
+                    // Compute only necessary hashes
+
+                    header.Header.UpdateHash(_serializier, _crypto);
+
+                    if (needHeader)
+                    {
+                        notFoundHeaders.Add(header.Header);
+                    }
+                    else
+                    {
+                        if (needBlock)
+                        {
+                            notFoundBlocks.Add(header.Header.Hash);
+                        }
+                    }
+                }
+            }
+
+            // Add only if is not in the chain
+
+            _blockchain.AddBlockHeaders(notFoundHeaders);
+
+            // Split on 500
+
+            for (int x = 0, count = notFoundBlocks.Count; x < count; x += 500)
             {
-                _logger.LogInformation($"The peer has {sender.Version.CurrentBlockIndex + 1} blocks but the current number of block headers is {_blockchain.LastBlockHeader.Index + 1}.");
-                await sender.Send(new GetBlockHeadersMessage(_blockchain.LastBlockHeader.Hash));
+                await sender.Send(new GetDataMessage(InventoryType.Block, notFoundBlocks.Skip(x).Take(500)));
             }
+
+            // Already sent on VerAck
+
+            //if (_blockchain.LastBlockHeader.Index < sender.Version.CurrentBlockIndex)
```

Comment:
Once again, without this condition, only the first 2000 block headers/blocks will be synced.